File: smoke_tests/ruby/resource_leak.rb
# Resource Leak vulnerabilities in Ruby
require 'socket'
require 'net/http'

class ResourceLeakController < ApplicationController
  # Test 1: File not closed
  def read_file_unsafe
    path = params[:path]
    # VULNERABLE: File not closed on exception
    file = File.open(path, 'r')
    content = file.read
    # file.close might not be reached
    render plain: content
  end

  # Test 2: Socket not closed
  def connect_socket
    host = params[:host]
    port = params[:port].to_i
    # VULNERABLE: Socket leak
    socket = TCPSocket.new(host, port)
    data = socket.gets
    render plain: data
    # socket never closed
  end

  # Test 3: Database connection leak
  def query_database
    conn = PG.connect(dbname: 'mydb')
    # VULNERABLE: Connection not closed
    result = conn.exec('SELECT * FROM users')
    render json: result.to_a
    # conn never closed
  end

  # Test 4: Net::HTTP not properly closed
  def fetch_url_unsafe
    uri = URI(params[:url])
    # VULNERABLE: HTTP connection may leak
    http = Net::HTTP.new(uri.host, uri.port)
    http.start
    response = http.get(uri.path)
    render plain: response.body
    # http never finished
  end

  # Test 5: Early return leak
  def conditional_read
    path = params[:path]
    condition = params[:condition]
    file = File.open(path, 'r')
    if condition == 'skip'
      # VULNERABLE: Early return without close
      return head(:ok)
    end
    content = file.read
    file.close
    render plain: content
  end

  # Test 6: Exception path leak
  def process_file_unsafe
    file = File.open('/tmp/data.txt', 'r')
    begin
      process_content(file.read)
    rescue StandardError
      # VULNERABLE: File not closed in rescue
      raise
    end
    file.close
    head :ok
  end

  # Test 7: Tempfile not cleaned
  def create_temp
    # VULNERABLE: Tempfile not unlinked
    temp = Tempfile.new('data')
    temp.write(params[:data])
    temp.close
    # temp.unlink not called
    render plain: temp.path
  end

  # Test 8: Multiple resources not closed
  def multiple_files
    file1 = File.open('/tmp/a.txt', 'r')
    # VULNERABLE: If this fails, file1 leaks
    file2 = File.open('/tmp/b.txt', 'r')
    content = file1.read + file2.read
    file1.close
    file2.close
    render plain: content
  end

  # Test 9: Dir not closed
  def list_directory
    path = params[:path]
    # VULNERABLE: Dir not closed
    dir = Dir.open(path)
    entries = dir.entries
    render json: entries
    # dir.close never called
  end

  # Test 10: IO.popen not closed
  def run_command
    command = params[:cmd]
    # VULNERABLE: IO not closed
    io = IO.popen(command)
    output = io.read
    render plain: output
    # io.close never called
  end

  # Test 11: StringIO in loop
  def process_items
    params[:items].each do |item|
      # VULNERABLE: Many StringIO objects created
      io = StringIO.new
      io.write(item)
      # io never closed
    end
    head :ok
  end

  # Test 12: Gzip stream leak
  def read_gzip
    path = params[:path]
    gz = Zlib::GzipReader.open(path)
    content = gz.read
    # VULNERABLE: gz.close never called
    render plain: content
  end

  private

  def process_content(content)
    raise 'Error' if content.empty?
  end
end

File: smoke_tests/php/resource_leak.php
<?php
// Resource Leak vulnerabilities in PHP

// Test 1: File not closed
function read_file_unsafe() {
    $path = $_GET['path'];
    // VULNERABLE: File handle not closed on exception
    $f = fopen($path, 'r');
    $content = fread($f, filesize($path));
    // fclose($f) might not be reached
    return $content;
}

// Test 2: Socket not closed
function connect_socket() {
    $host = $_GET['host'];
    $port = (int)$_GET['port'];
    // VULNERABLE: Socket leak
    $socket = fsockopen($host, $port);
    $data = fgets($socket);
    echo $data;
    // fclose($socket) never called
}

// Test 3: Database connection not closed
function query_database() {
    $pdo = new PDO('mysql:host=localhost;dbname=test', 'root', 'password');
    // VULNERABLE: Connection not closed
    $stmt = $pdo->query("SELECT * FROM users");
    return $stmt->fetchAll();
    // $pdo = null; not called
}

// Test 4: cURL not closed
function fetch_url() {
    $url = $_GET['url'];
    $ch = curl_init($url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    $result = curl_exec($ch);
    // VULNERABLE: curl_close($ch) never called
    return $result;
}

// Test 5: Early return leak
function conditional_read() {
    $path = $_GET['path'];
    $condition = $_GET['condition'];

    $f = fopen($path, 'r');
    if ($condition === 'skip') {
        // VULNERABLE: Early return without close
        return;
    }
    $content = fread($f, 1024);
    fclose($f);
    return $content;
}

// Test 6: Exception path leak
function process_file() {
    $f = fopen('/tmp/data.txt', 'r');
    try {
        process_content(fread($f, 1024));
    } catch (Exception $e) {
        // VULNERABLE: File not closed in catch
        throw $e;
    }
    fclose($f);
}

// Test 7: Multiple handles not all closed
function multiple_files() {
    $f1 = fopen('/tmp/a.txt', 'r');
    // VULNERABLE: If this fails, f1 leaks
    $f2 = fopen('/tmp/b.txt', 'r');

    $content = fread($f1, 1024) . fread($f2, 1024);

    fclose($f1);
    fclose($f2);
    return $content;
}

// Test 8: Directory handle not closed
function list_directory() {
    $path = $_GET['path'];
    // VULNERABLE: Dir handle not closed
    $dir = opendir($path);
    $entries = [];
    while (($entry = readdir($dir)) !== false) {
        $entries[] = $entry;
    }
    // closedir($dir) never called
    return $entries;
}

// Test 9: proc_open not closed
function run_command() {
    $cmd = $_GET['cmd'];
    $descriptors = [
        0 => ['pipe', 'r'],
        1 => ['pipe', 'w'],
        2 => ['pipe', 'w']
    ];
    // VULNERABLE: Process not closed
    $process = proc_open($cmd, $descriptors, $pipes);
    $output = stream_get_contents($pipes[1]);
    // proc_close($process) never called
    return $output;
}

// Test 10: GD image not destroyed
function process_image() {
    $path = $_GET['path'];
    // VULNERABLE: Image resource not destroyed
    $img = imagecreatefromjpeg($path);
    $width = imagesx($img);
    // imagedestroy($img) never called
    return $width;
}

// Test 11: XMLReader not closed
function read_xml() {
    $path = $_GET['path'];
    $reader = new XMLReader();
    $reader->open($path);
    // VULNERABLE: Reader not closed
    while ($reader->read()) {
        echo $reader->name;
    }
    // $reader->close() never called
}

// Test 12: Stream filter leak
function filtered_read() {
    $path = $_GET['path'];
    $f = fopen($path, 'r');
    $filter = stream_filter_append($f, 'string.toupper');
    $content = fread($f, 1024);
    // VULNERABLE: Filter and handle not cleaned up
    return $content;
}

function process_content($content) {
    if (empty($content)) {
        throw new Exception('Empty content');
    }
}
?>
